CosmosApiMessagesChangeFeed/handler.ts

```diff
@@ -0,0 +1,136 @@
+import { TableClient, TableInsertEntityHeaders } from "@azure/data-tables";
+import { BlobService } from "azure-storage";
+
+import { flow, pipe } from "fp-ts/lib/function";
+import * as T from "fp-ts/lib/Task";
+import * as TE from "fp-ts/lib/TaskEither";
+import * as E from "fp-ts/lib/Either";
+import * as RA from "fp-ts/ReadonlyArray";
+
+import {
+  MessageModel,
+  RetrievedMessage
+} from "@pagopa/io-functions-commons/dist/src/models/message";
+import { sequenceS } from "fp-ts/lib/Apply";
+import * as KP from "../utils/kafka/KafkaProducerCompact";
+import { publish } from "../utils/publish";
+import { IBulkOperationResult } from "../utils/bulkOperationResult";
+
+const CHUNK_SIZE = 15;
+
+/**
+ * Store errors in errorStorage table
+ *
+ * @param errorStorage
+ * @returns
+ */
+const storeMessageErrors = (errorStorage: TableClient) => (
+  storableErrors: ReadonlyArray<Error>
+): ReadonlyArray<TE.TaskEither<Error, TableInsertEntityHeaders>> =>
+  storableErrors.map(es =>
+    TE.tryCatch(
+      () =>
+        errorStorage.createEntity({
+          message: es.message,
+          name: "Message Error",
+          partitionKey: `${new Date().getMonth() + 1}`,
+          rowKey: `${Date.now()}`
+        }),
+      E.toError
+    )
+  );
+
+/**
+ * Retrieve a message content from blob storage and enrich message
+ */
+const enrichMessageContent = (
+  messageModel: MessageModel,
+  blobService: BlobService,
+  message: RetrievedMessage
+): TE.TaskEither<Error, RetrievedMessage> =>
+  pipe(
+    messageModel.getContentFromBlob(blobService, message.id),
+    TE.chain(TE.fromOption(() => Error(`Blob not found`))),
+    TE.mapLeft(err => Error(`Message ${message.id}: ${err.message}`)),
+    TE.map(content => ({
+      ...message,
+      content,
+      kind: "IRetrievedMessageWithContent"
+    }))
+  );
+
+/**
+ * Enrich messages with content, retrieved from blob storage, if exists
+ *
+ */
+export const enrichMessagesContent = (
+  messageModel: MessageModel,
+  mesageContentChunkSize: number,
+  blobService: BlobService,
+  errorStorage: TableClient
+) => (
+  messages: ReadonlyArray<RetrievedMessage>
+): T.Task<ReadonlyArray<RetrievedMessage>> =>
+  pipe(
+    messages,
+    // split execution in chunks of 'mesageContentChunkSize'
+    RA.chunksOf(mesageContentChunkSize),
+    RA.map(
+      flow(
+        RA.map(m =>
+          m.isPending === false
+            ? enrichMessageContent(messageModel, blobService, m)
+            : TE.of(m)
+        ),
+        // call task in parallel
+        RA.sequence(T.ApplicativePar)
+      )
+    ),
+    // call chunk tasks sequentially
+    RA.sequence(T.ApplicativeSeq),
+    T.map(RA.flatten),
+    T.map(mess => ({ errors: RA.lefts(mess), rights: RA.rights(mess) })),
+    TE.fromTask,
+    TE.chain(({ errors, rights }) =>
+      pipe(
+        errors,
+        storeMessageErrors(errorStorage),
+        RA.sequence(TE.ApplicativeSeq),
+        TE.bimap(
```

Comment:
Could be useful to throw an error if storage errors fails to retry the function and do not lose the messages

Reply:
Thanks, done in [4397783](https://github.com/pagopa/io-functions-elt/pull/39/commits/439778311e70c740a37991382f0eb23dc3fdcfb0)